ouroboros-network/test/Test/Ouroboros/Network/ShrinkCarefully.hs

```diff
@@ -0,0 +1,26 @@
+{- This module enables us to test that shrinking functions do not
+   shrink a value to itself. To use, define a property for your type
+   of interest like this:
+
+   prop_shrinkCarefully_T :: ShrinkCarefully T -> Property
+   prop_shrinkCarefully_T = prop_shrinkCarefully
+-}
+
+module Test.Ouroboros.Network.ShrinkCarefully where
+
+import Data.List
+import Text.Pretty.Simple
+import Test.QuickCheck
+
+newtype ShrinkCarefully a = ShrinkCarefully a
+  deriving (Eq,Show)
+
+instance (Eq a, Arbitrary a) => Arbitrary (ShrinkCarefully a) where
+  arbitrary = ShrinkCarefully <$> arbitrary
+  shrink (ShrinkCarefully a) = ShrinkCarefully <$> delete a (shrink a)
+
+prop_shrinkCarefully :: (Arbitrary a, Eq a, Show a) => ShrinkCarefully a -> Property
+prop_shrinkCarefully (ShrinkCarefully e) =
+  whenFail (pPrint e) $
+  counterexample (show $ map (==e) (shrink e)) $
+  e `notElem` shrink e
```

Comment:
That will certainly work. On the plus side, it avoids code duplication, and helps ensure that the error message and the actual property remain consistent if the property is modified in the future. On the minus side, it means you have to read two lines rather than one to understand what the property is actually testing. I think it's a bit swings and roundabouts.

Reply:
It's probably just because I am used to spot sharing opportunities the former is easier to read for me (does not trigger the sharing opportunity trigger).